docs/design/persistent-volume-provisioning.md

```diff
@@ -0,0 +1,140 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<h1>*** PLEASE NOTE: This document applies to the HEAD of the source
+tree only. If you are using a released version of Kubernetes, you almost
+certainly want the docs that go with that version.</h1>
+
+<strong>Documentation for specific releases can be found at
+[releases.k8s.io](http://releases.k8s.io).</strong>
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+# Persistent Volume Provisioning
+
+This document proposes a model for dynamically provisioning [Persistent Volumes](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/persistent-volumes.md)
+
+### tl;dr
+
+One new API kind:
+
+A `PersistentVolumeController` (PVC) is a storage resource provisioned by an administrator.  PVCs do not have a namespace.  Just as a `ReplicationController` maintains a number of replicas of a pod, a `PersistentVolumeController` maintains a minimum number of replicas of a `PersistentVolume`.  A PVC creates new volumes from a template up to a maximum replica count in increments of the minimum replica count.  A well-provisioned cluster will have many PVCs.
+
+One new system component:
+
+`PersistentVolumeControllerManager` is a singleton control loop running in master that manages all PVControllers in the system.  The PVCM reconciles the current supply of available PersistentVolumes in the system with the desired levels according to the PVControllers.   This process is similar to the ```ReplicationManager``` that manages ReplicationControllers.
+
+Three new volume plugin interfaces:
+
+* Recycler -- knows how to scrub a volume clean so it can become available again as a resource
+* Creator -- create new instances of a PV from a template.
+* Deleter -- deletes instances of a PV and allows the plugin to determine how to remove it from the underlying infrastructure
+
+Volume plugins can implement any applicable interfaces.  Each plugin will document its own support for dynamic provisioning.
+
+
+### Goals
+
+* Allow administrators to describe minimum and maximum storage levels comprised of many kinds of PersistentVolumes
+* Allow the dynamic creation and reclamation of persistent volumes (to the fullest extent for each type of storage provider)
```

Comment:
I think the key distinction is that the in tree controller would only
support a few obviously core types (like a cloud provider).  A custom
controller would not be in tree, or compiled with the manager, and we would
disable it.

On Thu, Aug 20, 2015 at 8:46 PM, Mark Turansky <notifications@github.com>
wrote:

> In docs/design/persistent-volume-provisioning.md
> <https://github.com/kubernetes/kubernetes/pull/6773#discussion_r37597422>:
>
> > +
> > +`PersistentVolumeControllerManager` is a singleton control loop running in master that manages all PVControllers in the system.  The PVCM reconciles the current supply of available PersistentVolumes in the system with the desired levels according to the PVControllers.   This process is similar to the ```ReplicationManager``` that manages ReplicationControllers.
> > +
> > +Three new volume plugin interfaces:
> > +
> > +* Recycler -- knows how to scrub a volume clean so it can become available again as a resource
> > +* Creator -- create new instances of a PV from a template.
> > +* Deleter -- deletes instances of a PV and allows the plugin to determine how to remove it from the underlying infrastructure
> > +
> > +Volume plugins can implement any applicable interfaces.  Each plugin will document its own support for dynamic provisioning.
> > +
> > +
> > +### Goals
> > +
> > +* Allow administrators to describe minimum and maximum storage levels comprised of many kinds of PersistentVolumes
> > +* Allow the dynamic creation and reclamation of persistent volumes (to the fullest extent for each type of storage provider)
>
> @smarterclayton <https://github.com/smarterclayton> yes, this works much
> more easily for volumes w/ APIs.
>
> I was thinking about making the strategy for creation be plugin-based.
> Perhaps the strategies are compiled in the /plugins package and you choose
> one by name via config (with a sensible default, of course).
>
> If we made creation pluggable, @pmorie <https://github.com/pmorie> can
> run his pod to create a volume or Go code can call a provider's API.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/6773/files#r37597422>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift